### ISSUE TITLE
is stuff like this really necessary?

### ISSUE THREAD

**Issue by computron** [materialsintelligence/propnet#85]
https://github.com/materialsintelligence/propnet/blob/master/propnet/models/EmpiricalBandGapSetyawan.py

**Comment by computron** [comment]
i.e. can you just have the YAML file and not the .py file

Or maybe put the YAML as a block string in the .py file

The double-ness of everything is annoying

**Comment by computron** [comment]
e.g. if the class in the .py file had a special attribute or method called "yaml_rep()" or something that had the YAML as string in block format in Python. 

- all info would be in the .py file (not separated between .py and .yaml)
- you don't need correspondence between filenames in .py and .yaml

**Comment by mkhorton** [comment]
It started off as all .py files, but the .py files were getting very messy (and, more importantly, *inconsistent* with each other), so the decision was to separate data into .yaml and logic into .py -- many simple models don't have any extra logic though, so they're .yaml only.

The .py stubs are mainly there for historical reasons (no one's written the code such that they're not necessary), but do have the bonus of helping with IDE code completion.

**Comment by mkhorton** [comment]
One point to mention about the .yaml file is that the footer (everything after the `---`) is valid Markdown, and this forms the description/documentation of the Model. This is a de facto format used in a lot of places, where there's a yaml front matter, followed by `---` and then a Markdown body.

The main advantage of this is that you can write a long description of how the model works quite nicely in a text editor and it remains readable to edit afterwards.

**Comment by computron** [comment]
What I was suggesting as a single .py file that looks something like this:

```python
from propnet.core.models import AbstractModel

yaml_data = """---
title: Estimate experimental band gaps from DFT (PBE) calculations
tags: [electronic]
references: ["doi:10.1021/co200012w"]
symbol_mapping: {
  E_g_PBE: band_gap_pbe,
  E_g_expt: band_gap
}
connections:
- inputs: E_g_expt
  outputs: [E_g_PBE]
equations: [E_g_expt - 1.348*E_g_pbe + 0.913]
---
Band gaps estimated using Density Functional Theory (and the PBE exchange-correlation
functional) typically underestimate the true band gap. A linear least squares fit
can provide an estimate of the true band gap from the PBE gap (R² = 0.886 when tested
against 100 compounds with gaps between ~1 and ~12 eV).
"""

class EmpiricalBandGapSetyawan(AbstractModel): pass
# add some code here to connect it to the yaml_data string above
```

**Comment by mkhorton** [comment]
Oh ok, yeah we could definitely do that!

**Comment by mkhorton** [comment]
Actually I think that’d make a lot of sense

**Comment by montoyjh** [comment]
so I'm gonna try to tackle this over the next couple of days, and think it'll be most productive just to start breaking things and see what sticks, but I figure I'd share some thoughts from my conversation just now with @dmrdjenovich about this:

I'm gonna start by just putting everything in the py files, but I think it'd be worthwhile to try to achieve a meaningful serialization format in the longer term.  In my mind there are two paths forward:
* Keeping the current class structure (i. e. every model is a distinct subclass of AbstractModel).  The difficulty with creating a serialization of these is that I'm not sure it's possible to dynamically create a class (not a class instance, but an actual class) from something that's not python code, and if it is possible I'm not sure it's something we should do.
* Use a single class for models, and create a meaningful serialization format (e. g. using something like from_dict).  The challenge here is that the structure of model necessitates the attachment of *methods*, rather than simple attributes to the class.  We can definitely use something like `eval` to create methods that get assigned to class attributes (or lambify for strings, etc.), but assigning methods can get a little dicey and makes for weird looking code regardless.

**Comment by computron** [comment]
A few other options

(i) you could *split* the models into YAML-compatible and not YAML-compatible (Python). The YAML ones will just be of class YAMLModel, which would be a class that can load any YAML data and use that to create a working model. Thus all YAML-based models would be of class YAMLModel, and the Python-based models could all be different classes.
(ii) If you don't like all of the YAML models to be of a single class, you should be able to use metaprogramming / metaclasses to dynamically create classes based on an initial read of your YAML files. Then each YAML file can have its own class created at runtime with its own class name, etc.

Regardless, you would want the serialization format to just follow monty (?). For the YAML based ones you might have to overload to_dict / from_dict

**Comment by mkhorton** [comment]
I think this has largely been addressed by #111 so will close
